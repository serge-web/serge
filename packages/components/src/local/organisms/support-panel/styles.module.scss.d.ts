// This file is automatically generated.
// Please do not change this file!
interface CssExports {
  'action-tab': string;
  'active': string;
  'content': string;
  'new-order-input': string;
  'order-group': string;
  'panel': string;
  'resize-indicator-container': string;
  'resize-indicator-icon': string;
  'root': string;
  'secondary-action-tab': string;
  'select-template-title': string;
  'tab-panel': string;
}
export const cssExports: CssExports;
export default cssExports;
