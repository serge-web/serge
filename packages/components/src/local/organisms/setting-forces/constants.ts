export const PLATFORM_ITEM = 'PLATFORM_ITEM'
export const ASSET_ITEM = 'ASSET_ITEM'
