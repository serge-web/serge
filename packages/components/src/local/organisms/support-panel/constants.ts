import { CSSProperties } from 'react'

export const TAB_MY_ORDERS = 'my_orders'
export const TAB_ADJUDICATE = 'adjudication'
export const TAB_MY_FORCE = 'my_force'
export const TAB_OPP_FOR = 'opp_force'

export const MIN_PANEL_WIDTH = 330
export const MAX_PANEL_WIDTH = '80%'
export const MIN_PANEL_HEIGHT = '100%'
export const MAX_PANEL_HEIGHT = '100%'

export const PANEL_STYLES: CSSProperties = {
  display: 'flex',
  position: 'unset'
}

export const DEFAULT_SIZE = {
  x: 0,
  y: 0,
  width: MIN_PANEL_WIDTH,
  height: '100%'
}

export const LRU_CACHE_OPTION = {
  max: 500,
  maxSize: 5000,
  ttl: 1000 * 60 * 60,
  maxEntrySize: 5000,
  sizeCalculation: (): number => 1
}
