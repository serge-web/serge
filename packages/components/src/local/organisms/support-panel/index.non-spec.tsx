/* global it expect */
import { ChannelPlanning, ParticipantTemplate, TemplateBody } from '@serge/custom-types'
import { checkV3ParticipantStates } from '@serge/helpers'
import { P9Mock, planningMessages, planningMessageTemplatesMock } from '@serge/mocks'
import { noop } from 'lodash'
import React from 'react'
import renderer from 'react-test-renderer'
import SupportPanel from './index'

const planningChannel = P9Mock.data.channels.channels[0] as ChannelPlanning
const forces = P9Mock.data.forces.forces
const blueForce = forces[1]
const blueRole = blueForce.roles[0]
const thisPart = checkV3ParticipantStates(planningChannel, blueForce.uniqid, blueRole.roleId, false)

const myTemplateIds = thisPart.templatesIDs
const myTemplates = planningMessageTemplatesMock.filter((value: TemplateBody): any =>
  myTemplateIds.find((id: ParticipantTemplate) => id._id === value._id)
)
const platformTypes = (P9Mock.data.platformTypes && P9Mock.data.platformTypes.platformTypes) || []

describe('Support Panel component: ', () => {
  it('renders component correctly', () => {
    const tree = renderer
      .create(<SupportPanel
        platformTypes={platformTypes}
        allForces={P9Mock.data.forces.forces}
        gameDate={P9Mock.data.overview.gameDate}
        currentWargame={P9Mock.currentWargame || ''}
        selectedAssets={[]}
        setSelectedAssets={noop}
        selectedOrders={[]}
        setSelectedOrders={noop}
        currentTurn={P9Mock.gameTurn}
        messages={planningMessages}
        channel={planningChannel}
        onUnread={noop}
        onReadAll={noop}
        onRead={noop}
        templates={myTemplates}
        adjudicationTemplate={planningMessageTemplatesMock[0]}
        activityTimeChanel={noop}
        dispatch={noop}
        saveMessage={(): any => noop}
        saveNewActivityTimeMessage={(): any => noop}
        selectedForce={blueForce}
        selectedRoleId={blueRole.roleId}
        selectedRoleName={blueRole.name}
        isUmpire={false}
        setOpForcesForParent={noop}
        setOwnForcesForParent={noop}
        allOppAssets={[]}
        allOwnAssets={[]}
        activities={[]}
      />)
      .toJSON()
    expect(tree).toMatchSnapshot()
  })
})
