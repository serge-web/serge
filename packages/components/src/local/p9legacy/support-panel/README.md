# Support Panel documentation

This is an example readme file for the 'Support Panel' component.

## UI Example

<!-- STORY -->

## Code example

<!-- SOURCE -->

## Props

<!-- PROPS -->

## Usage

### ES6 Import

```js
import SupportPanel from "./support-panel";
```

### CommonJS

```js
const SupportPanel = require("./support-panel");
```
