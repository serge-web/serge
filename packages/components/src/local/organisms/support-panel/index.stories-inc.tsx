import { INFO_MESSAGE_CLIPPED, INTERACTION_MESSAGE, Phase, PLANNING_MESSAGE } from '@serge/config'
import { ChannelPlanning, ForceData, MessageDetails, MessageInteraction, MessagePlanning, Role } from '@serge/custom-types'
import { forceColors, incrementGameTime, platformIcons } from '@serge/helpers'
import { P9BMock, planningMessages as planningChannelMessages, turnPeriod } from '@serge/mocks'
import { withKnobs } from '@storybook/addon-knobs'
import { Story } from '@storybook/react/types-6-0'
import { cloneDeep, noop } from 'lodash'
import moment from 'moment'
import React from 'react'
import { getOppAssets, getOwnAssets } from '../planning-assets/helpers/collate-assets'
import { TAB_ADJUDICATE, TAB_MY_ORDERS } from './constants'
import SupportPanel from './index'
import docs from './README.md'
import SupportPanelProps from './types/props'

const wrapper: React.FC = (storyFn: any) => <div style={{ height: '600px' }}>{storyFn()}</div>

const planningChannel = P9BMock.data.channels.channels[0] as ChannelPlanning
const forces = P9BMock.data.forces.forces
const templates = P9BMock.data.templates ? P9BMock.data.templates.templates : []
const gameDate = P9BMock.data.overview.gameDate
const gameTime = moment(gameDate).valueOf()

const allRoles: string[] = []
forces.forEach((force: ForceData) => {
  force.roles.forEach((role: Role) => {
    allRoles.push(force.uniqid + ' ~ ' + role.roleId)
  })
})

const nonInfoMessage = planningChannelMessages.filter((msg) => msg.messageType !== INFO_MESSAGE_CLIPPED) as Array<MessagePlanning | MessageInteraction>
const planningMessages = nonInfoMessage.filter((msg: MessagePlanning | MessageInteraction) => msg.messageType === PLANNING_MESSAGE) as Array<MessagePlanning>
const interactionMessages = nonInfoMessage.filter((msg: MessagePlanning | MessageInteraction) => msg.messageType === INTERACTION_MESSAGE) as Array<MessageInteraction>
const attributeTypes = P9BMock.data.attributeTypes ? P9BMock.data.attributeTypes.attributes : []

const handler = (contact: any): void => {
  console.log('handling', contact)
}

export default {
  title: 'local/organisms/SupportPanel',
  component: SupportPanel,
  decorators: [withKnobs, wrapper],
  parameters: {
    readme: {
      // Show readme before story
      content: docs
    },
    options: {
      // We have no addons enabled in this story, so the addon panel should be hidden
      showPanel: true
    },
    controls: {
      expanded: true
    }
  },
  argTypes: {
    selectedRoleName: {
      name: 'View as',
      defaultValue: allRoles[1],
      options: allRoles,
      control: {
        type: 'select'
      }
    }
  }
}

const platformTypes = (P9BMock.data.platformTypes && P9BMock.data.platformTypes.platformTypes) || []
const activities = P9BMock.data.activities ? P9BMock.data.activities.activities : []

// produce the own and opp assets for this player force
const forceCols = forceColors(forces)
const platIcons = platformIcons(platformTypes)
const own = getOwnAssets(forces, forceCols, platIcons, forces[1], platformTypes, attributeTypes, gameTime)
const opp = getOppAssets(forces, forceCols, platIcons, forces[1], platformTypes, attributeTypes, gameTime)

const Template: Story<SupportPanelProps> = (args) => {
  const roleStr: string = args.selectedRoleName
  const initialTab = args.initialTab
  const planningMessages = args.planningMessages
  const interactionMessages = args.interactionMessages

  // separate out the two elements of the combined role
  const ind = roleStr.indexOf(' ~ ')
  const forceStr = roleStr.substring(0, ind)
  const role = roleStr.substring(ind + 3)

  const saveMessage = (dbName: string, details: MessageDetails, message: any) => {
    return async (): Promise<void> => {
      console.log('dbName: ', dbName, ', details: ', details, ', message: ', message)
    }
  }

  const force = forces.find((value: ForceData) => value.uniqid === forceStr)
  if (!force) {
    throw Error('can\'t find force')
  }
  const roleVal = force.roles.find((roleVal: Role) => roleVal.roleId === role)
  if (!roleVal) {
    throw Error('can\'t find role')
  }

  return <SupportPanel
    platformTypes={platformTypes}
    planningMessages={planningMessages}
    interactionMessages={interactionMessages}
    forcePlanningActivities={activities}
    onReadAll={noop}
    selectedAssets={[]}
    setSelectedAssets={noop}
    selectedOrders={[]}
    attributeTypes={attributeTypes}
    handleAdjudication={handler}
    setSelectedOrders={noop} onUnread={noop}
    onRead={noop}
    phase={Phase.Planning}
    channel={planningChannel}
    allTemplates={templates}
    allPeriods={turnPeriod}
    adjudicationTemplate={templates[0]}
    activityTimeChanel={noop}
    allForces={P9BMock.data.forces.forces}
    gameDate={P9BMock.data.overview.gameDate}
    gameTurnLength={P9BMock.data.overview.gameTurnTime}
    currentWargame={P9BMock.currentWargame || ''}
    currentTurn={P9BMock.gameTurn}
    dispatch={noop}
    saveMessage={saveMessage}
    saveNewActivityTimeMessage={() => (): void => { console.log('save activity') }}
    selectedRoleId={roleVal.roleId}
    selectedRoleName={roleVal.name}
    selectedForce={force}
    setOpForcesForParent={noop}
    setOwnForcesForParent={noop}
    allOppAssets={opp}
    allOwnAssets={own}
    initialTab={initialTab}
  />
}

// mangle some dates
const firstTurn = turnPeriod[0]
const startTime = moment.utc(firstTurn.gameDate).valueOf()
const secondTurn = turnPeriod[1]
const secondEnd = incrementGameTime(secondTurn.gameDate, secondTurn.gameTurnTime)
const secondEndTime = moment.utc(secondEnd).valueOf()
const secondStartTime = moment.utc(secondTurn.gameDate).valueOf()
const midPoint = secondStartTime + (secondEndTime - secondStartTime) / 2

const newPlans = cloneDeep(planningMessages) as MessagePlanning[]
const msgToMove1 = newPlans[4]
msgToMove1.message.startDate = moment.utc(midPoint).toISOString()
msgToMove1.message.endDate = moment.utc(midPoint + 500000).toISOString()
const msgToMove2 = newPlans[5]
msgToMove2.message.startDate = moment.utc(startTime - 100000).toISOString()
msgToMove2.message.endDate = moment.utc(midPoint + 500000).toISOString()

const newInter = cloneDeep(interactionMessages) as MessageInteraction[]
const interToMove1 = newInter[2]
const intertoMove2 = newInter[3]
if (interToMove1.details.interaction) {
  interToMove1.details.interaction = {
    ...interToMove1.details.interaction,
    startTime: moment.utc(midPoint).toISOString(),
    endTime: moment.utc(midPoint + 500000).toISOString()
  }
}
if (intertoMove2.details.interaction) {
  intertoMove2.details.interaction = {
    ...intertoMove2.details.interaction,
    startTime: moment.utc(startTime - 100000).toISOString(),
    endTime: moment.utc(midPoint + 500000).toISOString()
  }
}

export const Default = Template.bind({})
Default.args = {
  planningMessages: planningMessages,
  interactionMessages: newInter
}

export const OrdersTab = Template.bind({})
OrdersTab.args = {
  initialTab: TAB_MY_ORDERS,
  planningMessages: newPlans,
  interactionMessages: newInter
}

export const AdjudicationTab = Template.bind({})
AdjudicationTab.args = {
  initialTab: TAB_ADJUDICATE,
  planningMessages: newPlans,
  interactionMessages: newInter,
  selectedRoleName: allRoles[1]
}

// console.log('fixing data')
// forces.forEach((force) => {
//   if (force.assets) {
//     force.assets.forEach((asset) => {
//       if (asset.platformTypeId.includes('istar')) {
//         if (asset.attributes) {
//           asset.attributes.a_Search_Rate = 110
//           console.log('updated')
//         } else {
//           console.warn('found asset without attributes')
//         }
//       }
//     })
//   }
// })
// console.log('updated', forces)
