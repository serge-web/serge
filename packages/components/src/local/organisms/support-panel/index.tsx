import Slide from '@material-ui/core/Slide'
import MoreVert from '@material-ui/icons/MoreVert'
import { ADJUDICATION_PHASE, MESSAGE_SENT_INTERACTION } from '@serge/config'
import { MessageDetails, MessageInteraction, MessagePlanning, MessageSentInteraction, MessageStructure, PerForcePlanningActivitySet, PlannedActivityGeometry, PlannedProps, PlanningMessageStructureCore } from '@serge/custom-types'
import { forceColors, ForceStyle, incrementGameTime, platformIcons, PlatformStyle } from '@serge/helpers'
import cx from 'classnames'
import { noop } from 'lodash'
import moment from 'moment'
import React, { createContext, useContext, useEffect, useMemo, useState } from 'react'
import { Rnd } from 'react-rnd'
import NewMessage from '../../form-elements/new-message'
import AdjudicationMessagesList from '../adjudication-messages-list'
import { AdjudicationRow } from '../adjudication-messages-list/types/props'
import PlanningAssets from '../planning-assets'
import { AssetRow } from '../planning-assets/types/props'
import PlanningMessagesList from '../planning-messages-list'
import { collapseLocation, expandLocation } from '../planning-messages-list/helpers/collapse-location'
import { OrderRow } from '../planning-messages-list/types/props'
import { DEFAULT_SIZE, MAX_PANEL_HEIGHT, MAX_PANEL_WIDTH, MIN_PANEL_HEIGHT, MIN_PANEL_WIDTH, PANEL_STYLES, TABS, TAB_ADJUDICATE, TAB_MY_ORDERS } from './constants'
import { customiseActivities } from './helpers/customise-activities'
import { customiseAssets } from './helpers/customise-assets'
import { customiseLiveOrders } from './helpers/customise-live-orders'
import { customiseLocation } from './helpers/customise-location'
import TurnFilter, { SHOW_ALL_TURNS } from './helpers/TurnFilter'
import { updateLocationTimings } from './helpers/update-location-timings'
import styles from './styles.module.scss'
import PropTypes, { PanelActionTabsProps, SupportPanelContextInterface, TabPanelProps } from './types/props'
export const SupportPanelContext = createContext<SupportPanelContextInterface>({ selectedAssets: [], setCurrentAssets: noop, setCurrentOrders: noop })

export const SupportPanel: React.FC<PropTypes> = ({
  platformTypes,
  planningMessages,
  interactionMessages,
  turnPresentation,
  onRead,
  onUnread,
  onReadAll,
  channel,
  allTemplates,
  adjudicationTemplate,
  mapPostBack,
  saveMessage,
  saveNewActivityTimeMessage,
  saveCachedNewMessageValue,
  getCachedNewMessagevalue,
  clearCachedNewMessage,
  selectedForce,
  selectedRoleId,
  selectedRoleName,
  allForces,
  gameDate,
  gameTurnLength: gameTurnTime,
  currentTurn,
  phase,
  currentWargame,
  setSelectedAssets,
  selectedOrders,
  setSelectedOrders,
  setOpForcesForParent,
  setOwnForcesForParent,
  allOppAssets,
  allOwnAssets,
  onPanelWidthChange,
  draftMessage,
  onCancelDraftMessage,
  forcePlanningActivities,
  editLocation,
  attributeTypes,
  handleAdjudication
}) => {
  const umpireInAdjudication = selectedForce.umpire && (phase === ADJUDICATION_PHASE)
  const [activeTab, setActiveTab] = useState<string>(umpireInAdjudication ? TABS[3] : TABS[0])
  const [isShowPanel, setShowPanel] = useState<boolean>(true)
  const [forceCols] = useState<ForceStyle[]>(forceColors(allForces))
  const [platIcons] = useState<PlatformStyle[]>(platformIcons(platformTypes))

  const [gameTurnEndDate, setGameTurnEndDate] = useState<string>('')

  const [selectedOwnAssets, setSelectedOwnAssets] = useState<AssetRow[]>([])
  const [selectedOpAssets, setSelectedOpAssets] = useState<AssetRow[]>([])
  const [filteredPlanningMessages, setFilteredPlanningMessages] = useState<MessagePlanning[]>([])
  const [filteredInteractionMessages, setFilteredInteractionMessages] = useState<MessageInteraction[]>([])
  const [turnFilter, setTurnFilter] = useState<number>(-1)
  const [localDraftMessage, setLocalDraftMessage] = useState<MessagePlanning | undefined>(undefined)
  const [activitiesForThisForce, setActivitiesForThisForce] = useState<PerForcePlanningActivitySet | undefined>(undefined)
  const [pendingLocationData, setPendingLocationData] = useState<PlannedActivityGeometry[]>([])
  const { setCurrentOrders, setCurrentAssets } = useContext(SupportPanelContext)

  const onTabChange = (tab: string): void => {
    setShowPanel(activeTab !== tab || !isShowPanel)
    setActiveTab(tab)
  }

  useEffect(() => {
    setLocalDraftMessage(draftMessage)
  }, [draftMessage])

  useEffect(() => {
    if (gameDate !== '' && gameTurnTime) {
      const endDate = incrementGameTime(gameDate, gameTurnTime)
      console.log('calc end date', gameDate, gameTurnTime, endDate)
      setGameTurnEndDate(endDate)
    }
    setLocalDraftMessage(draftMessage)
  }, [gameDate, gameTurnTime])

  useEffect(() => {
    if (forcePlanningActivities) {
      const thisForce = forcePlanningActivities.find((act: PerForcePlanningActivitySet) => act.force === selectedForce.uniqid)
      setActivitiesForThisForce(thisForce)
    }
  }, [forcePlanningActivities, selectedForce])

  useEffect(() => {
    const filtered = turnFilter === SHOW_ALL_TURNS ? planningMessages : planningMessages.filter((msg) => msg.details.turnNumber === turnFilter)
    setFilteredPlanningMessages(filtered)
  }, [planningMessages, turnFilter])

  useEffect(() => {
    const filtered = turnFilter === SHOW_ALL_TURNS ? interactionMessages : interactionMessages.filter((msg) => msg.details.turnNumber === turnFilter)
    setFilteredInteractionMessages(filtered)
  }, [interactionMessages, turnFilter])

  const TabPanel = (props: TabPanelProps): React.ReactElement => {
    const { children, active, ...other } = props
    return (
      <div
        hidden={!active}
        {...other}
      >
        {children}
      </div>
    )
  }

  const TabPanelActions = ({ onChange, className }: PanelActionTabsProps): React.ReactElement => {
    return (
      <div className={cx(styles['action-tab'], className)}>
        <p onClick={(): void => onChange(TABS[0])} className={cx({ [styles.active]: activeTab === TABS[0] })}>My Force</p>
        <p onClick={(): void => onChange(TABS[1])} className={cx({ [styles.active]: activeTab === TABS[1] })}>Orders</p>
        <p onClick={(): void => onChange(TABS[2])} className={cx({ [styles.active]: activeTab === TABS[2] })}>OPFOR</p>
        {selectedForce.umpire && <p onClick={(): void => onChange(TABS[3])} className={cx({ [styles.active]: activeTab === TABS[3] })}>Adjudication</p>
        }
      </div>
    )
  }

  const onRender = (): void => {
    console.log('=> render')
  }

  useEffect(() => {
    const allSelectedAssets = selectedOwnAssets.concat(selectedOpAssets)
    const selectedAssetIDs = allSelectedAssets.map((row: AssetRow) => row.id)
    setSelectedAssets(selectedAssetIDs)
  }, [selectedOwnAssets, selectedOpAssets])

  useEffect(() => {
    // if there is a draft message, open the `my orders` tab
    if (draftMessage) {
      setActiveTab(TAB_MY_ORDERS)
    }
  }, [draftMessage])

  useEffect(() => {
    onPanelWidthChange && onPanelWidthChange(isShowPanel ? 330 : 0)
  }, [isShowPanel])

  const onVisibleRowsChange = (opFor: boolean, data: AssetRow[]): void => {
    if (opFor) {
      setOpForcesForParent(data)
    } else {
      setOwnForcesForParent(data)
    }
  }

  const cancelNewOrders = (): void => {
    onCancelDraftMessage && onCancelDraftMessage()
    // also the local document
    setLocalDraftMessage(undefined)
  }

  const postBack = (details: MessageDetails, message: any): void => {
    // do we have any pending geometry
    if (pendingLocationData.length > 0) {
      const plan = message as MessagePlanning
      console.log('injecting geometry', plan.message.location, pendingLocationData)
    }

    const activity: MessageSentInteraction = {
      aType: MESSAGE_SENT_INTERACTION
    }

    saveNewActivityTimeMessage(selectedRoleId, activity, currentWargame)

    // fix the location bits, if necessary
    const updatedDoc = localModifyForSave(message)
    console.log(message, updatedDoc)

    saveMessage(currentWargame, details, message)()
    // also clear local one
    setLocalDraftMessage(undefined)
  }

  const onSizeChange = (_: MouseEvent | TouchEvent, __: any, elementRef: HTMLElement): void => {
    onPanelWidthChange && onPanelWidthChange(elementRef.offsetWidth)
  }

  const onTurnFilterChange = (turn: number) => {
    setTurnFilter(turn)
  }

  const localCustomiseTemplate = (document: MessageStructure | undefined, schema: Record<string, any>): Record<string, any> => {
    // sort out which orders are currently "live"
    const turnStart = moment(gameDate)
    const turnEnd = moment(gameTurnEndDate)
    const liveOrders: MessagePlanning[] = planningMessages.filter((plan: MessagePlanning) => {
      const startDate = plan.message.startDate
      const endDate = plan.message.endDate
      if (startDate && endDate) {
        const startD = moment(startDate)
        const endD = moment(endDate)
        return startD.isBefore(turnEnd) && endD.isAfter(turnStart)
      } else {
        console.warn('Support panel. Orders start/end missing for', plan)
        return false
      }
    })

    // now modify the template
    const customisers: Array<{ (document: MessageStructure | undefined, schema: Record<string, any>): Record<string, any> }> = [
      (document, template) => customiseAssets(document, template, allOwnAssets, allOppAssets),
      (document, template) => customiseActivities(document, template, forcePlanningActivities || [], selectedForce),
      (document, template) => customiseLocation(document, template),
      (document, template) => customiseLiveOrders(document, template, liveOrders)
    ]

    let current: Record<string, any> = { ...schema }
    customisers.forEach((fn) => {
      current = fn(document, current)
    })
    return current
  }

  const editThisMessage = (docId: string): void => {
    const order = planningMessages.find((doc) => doc._id === docId)
    setLocalDraftMessage(order)
  }

  const assetsForOrders = (id?: string): string[] => {
    let res: string[] = []
    const plan = planningMessages.find((msg) => msg._id === id)
    if (plan) {
      const mine = plan.message.ownAssets || []
      const myIds = mine.map((val: { asset: string, number: number }): string => val.asset)
      const others = plan.message.otherAssets ? plan.message.otherAssets.map((val: { asset: string }): string => val.asset) : []
      res = myIds.concat(others)
    }
    return res
  }

  const onDetailPanelOpen = (rowData: OrderRow | AdjudicationRow) => {
    // if this is an orders item, or an adjudication, mark the relevant data
    // as 'current
    switch (activeTab) {
      case TAB_MY_ORDERS: {
        const order = rowData as OrderRow
        const plan = planningMessages.find((msg) => msg._id === order.id)
        if (plan) {
          const mine = plan.message.ownAssets || []
          const myIds = mine.map((val: { asset: string, number: number }): string => val.asset)
          const others = plan.message.otherAssets ? plan.message.otherAssets.map((val: { asset: string }): string => val.asset) : []
          const allIds = myIds.concat(others)
          setCurrentAssets(allIds)
          setCurrentOrders([plan._id])
        }
        break
      }
      case TAB_ADJUDICATE: {
        const adj = rowData as AdjudicationRow
        const doc = interactionMessages.find((doc) => doc._id === adj.id)
        if (doc) {
          const inter = doc.details.interaction
          if (inter) {
            // get the assets
            const assets1 = assetsForOrders(inter.orders1)
            const assets2 = assetsForOrders(inter.orders2)
            const allAssets = assets1.concat(assets2)
            setCurrentAssets(allAssets)
            if (inter.orders2) {
              setCurrentOrders([inter.orders1, inter.orders2])
            } else {
              setCurrentOrders([inter.orders1])
            }
          }
        }
      }
    }
  }

  const onDetailPanelClose = () => {
    setCurrentAssets([])
    setCurrentOrders([])
  }

  const storeNewLocation = (geoms: PlannedActivityGeometry[]): void => {
    console.log('storing new geometries', geoms)
    setPendingLocationData(geoms)
  }

  const localEditLocation = (geoms: PlannedActivityGeometry[]): void => {
    console.log('local edit location', geoms)
    setPendingLocationData(geoms)
    editLocation && editLocation(geoms, storeNewLocation)

    // if (message.message.location) {
    //   const localCallback = (newValue: unknown): void => {
    //     pendingLocationData.push(newValue as PlannedActivityGeometry[])
    //   }
    //   // pass the location data object
    //   editLocation && editLocation(message.message.location, localCallback)
    // }
  }

  const checkDisplay = (tab: string): string => {
    if (typeof tab === 'string' && tab === activeTab) {
      return 'block'
    }
    return 'none'
  }

  const onLocationEditorLoaded = (editorElm: HTMLDivElement) => {
    console.log('editorElm: ', editorElm)
  }

  const summariseLocations = (title: string, plans: PlannedActivityGeometry[]): void => {
    console.log('== ' + title + ' ==')
    console.table(plans.map((plan: PlannedActivityGeometry) => {
      const props = plan.geometry.properties as PlannedProps
      return {
        id: plan.uniqid,
        start: props.startDate,
        end: props.endDate
      }
    }))
  }

  const localModifyForSave = (document: Record<string, any>): Record<string, any>  => {
    const fixedLocation = expandLocation(document)
    const planDoc = fixedLocation as PlanningMessageStructureCore
    if (planDoc.location && planDoc.ownAssets) {
      const generalDoc = planDoc as any
      console.log('mod for save', planDoc, generalDoc.hiddenStore)
      const ownAssets = planDoc.ownAssets.map((item: {asset: string}) => item.asset)
      const updatedLocations = updateLocationTimings(planDoc.Reference, planDoc.location, ownAssets, allForces, planDoc.startDate, planDoc.endDate)
      summariseLocations('before', planDoc.location)
      summariseLocations('after', updatedLocations)
      planDoc.location = updatedLocations
    }
    return planDoc
  }

  const SlideComponent = useMemo(() => (
    <Slide direction="right" in={isShowPanel}>
      <div className={styles.panel}>
        <Rnd
          disableDragging
          style={PANEL_STYLES}
          default={DEFAULT_SIZE}
          minWidth={MIN_PANEL_WIDTH}
          maxWidth={MAX_PANEL_WIDTH}
          minHeight={MIN_PANEL_HEIGHT}
          maxHeight={MAX_PANEL_HEIGHT}
          onResize={onSizeChange}
        >
          <div className={styles.content}>
            <TabPanel className={styles['tab-panel']} value={TABS[0]} active={activeTab === TABS[0]}>
              {
                activeTab === TABS[0] &&
                <PlanningAssets
                  forceColors={forceCols}
                  assets={allOwnAssets}
                  attributeTypes={attributeTypes}
                  platformStyles={platIcons}
                  forces={allForces}
                  playerForce={selectedForce}
                  platformTypes={platformTypes}
                  render={onRender}
                  opFor={false}
                  onSelectionChange={setSelectedOwnAssets}
                  onVisibleRowsChange={(data): void => onVisibleRowsChange(false, data)}
                />
              }
            </TabPanel>
            {
              <div style={{ display: checkDisplay(TAB_MY_ORDERS) }} className={styles['order-group']}>
                <TurnFilter label='Show orders for turn:' currentTurn={currentTurn} value={turnFilter} onChange={onTurnFilterChange} />
                <PlanningMessagesList
                  messages={filteredPlanningMessages}
                  gameDate={gameDate}
                  gameTurnEndDate={gameTurnEndDate}
                  playerForceId={selectedForce.uniqid}
                  playerRoleId={selectedRoleId}
                  isUmpire={!!selectedForce.umpire}
                  turnPresentation={turnPresentation}
                  selectedForce={selectedForce}
                  selectedRoleName={selectedRoleName}
                  currentTurn={currentTurn}
                  hideForcesInChannel={false}
                  onRead={onRead}
                  onUnread={onUnread}
                  onMarkAllAsRead={onReadAll}
                  channel={channel}
                  allTemplates={allTemplates}
                  confirmCancel={true}
                  customiseTemplate={localCustomiseTemplate}
                  selectedOrders={selectedOrders}
                  setSelectedOrders={setSelectedOrders}
                  postBack={postBack}
                  turnFilter={turnFilter}
                  editLocation={editLocation}
                  modifyForSave={localModifyForSave}
                  forcePlanningActivities={forcePlanningActivities}
                  onDetailPanelOpen={onDetailPanelOpen}
                  onDetailPanelClose={onDetailPanelClose}
                  editThisMessage={editThisMessage}
                />
                {localDraftMessage && <NewMessage
                  orderableChannel={true}
                  privateMessage={!!selectedForce.umpire}
                  templates={allTemplates}
                  title={'New Orders'}
                  hideTemplateSelector={true}
                  saveCachedNewMessageValue={saveCachedNewMessageValue}
                  getCachedNewMessagevalue={getCachedNewMessagevalue}
                  clearCachedNewMessage={clearCachedNewMessage}
                  selectedRole={selectedRoleId}
                  selectedForce={selectedForce}
                  selectedRoleName={selectedRoleName}
                  confirmCancel={true}
                  onCancel={cancelNewOrders}
                  channel={channel}
                  currentTurn={currentTurn}
                  gameDate={gameDate}
                  postBack={postBack}
                  customiseTemplate={localCustomiseTemplate}
                  modifyForEdit={(document) => collapseLocation(document, activitiesForThisForce)}
                  modifyForSave={localModifyForSave}
                  draftMessage={localDraftMessage}
                  editCallback={localEditLocation}
                />}
              </div>
            }
            <TabPanel className={styles['tab-panel']} value={TABS[2]} active={activeTab === TABS[2]} >
              {activeTab === TABS[2] &&
                <PlanningAssets
                  forceColors={forceCols}
                  platformStyles={platIcons}
                  attributeTypes={attributeTypes}
                  assets={allOppAssets}
                  platformTypes={platformTypes}
                  forces={allForces}
                  playerForce={selectedForce}
                  render={onRender}
                  opFor={true}
                  onSelectionChange={setSelectedOpAssets}
                  onVisibleRowsChange={(data): void => onVisibleRowsChange(true, data)}
                />
              }
            </TabPanel>
            {activeTab === TABS[3] &&
              <div className={styles['order-group']}>
                <TurnFilter label='Show interactions for turn:' currentTurn={currentTurn} value={turnFilter} onChange={onTurnFilterChange} />
                <AdjudicationMessagesList
                  interactionMessages={filteredInteractionMessages}
                  planningMessages={filteredPlanningMessages}
                  forces={allForces}
                  gameDate={gameDate}
                  playerRoleId={selectedRoleId}
                  forceColors={forceCols}
                  onRead={onRead}
                  onUnread={onUnread}
                  onMarkAllAsRead={onReadAll}
                  mapPostBack={mapPostBack}
                  channel={channel}
                  template={adjudicationTemplate}
                  customiseTemplate={localCustomiseTemplate}
                  forcePlanningActivities={forcePlanningActivities}
                  turnFilter={turnFilter}
                  platformTypes={platformTypes}
                  onDetailPanelOpen={onDetailPanelOpen}
                  onDetailPanelClose={onDetailPanelClose}
                  handleAdjudication={handleAdjudication}
                  postBack={postBack}
                  onLocationEditorLoaded={onLocationEditorLoaded}
                />
              </div>
            }
            <div className={styles['resize-indicator-container']} >
              <div className={styles['resize-indicator-icon']} >
                <MoreVert fontSize='large' color='primary' style={{ marginLeft: 0 }} />
              </div>
            </div>
          </div>
        </Rnd>
        <TabPanelActions onChange={onTabChange} />
      </div>
    </Slide>
  ), [
    isShowPanel,
    activeTab,
    allForces,
    filteredPlanningMessages,
    filteredInteractionMessages,
    selectedRoleId,
    turnFilter,
    draftMessage,
    selectedOrders,
    platformTypes,
    planningMessages,
    localDraftMessage
  ]
  )

  return (
    <div className={styles.root}>
      {SlideComponent}
      <TabPanelActions onChange={onTabChange} className={styles['secondary-action-tab']} />
    </div>
  )
}

const areEqual = (prevProps: PropTypes, nextProps: PropTypes): boolean => JSON.stringify(prevProps) === JSON.stringify(nextProps)

export default React.memo(SupportPanel, areEqual)
