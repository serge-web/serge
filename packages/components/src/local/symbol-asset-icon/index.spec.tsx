/* global it expect */
import LRU from 'lru-cache'
import React from 'react'
import renderer from 'react-test-renderer'
import { LRU_CACHE_OPTION } from '../organisms/support-panel/constants'
import SymbolAssetIcon from './index'

it('renders correctly SymbolAssetIcon', () => {
  const tree = renderer
    .create(<SymbolAssetIcon sidc='' iconName='' assetsCache={new LRU(LRU_CACHE_OPTION)} />)
    .toJSON()
  expect(tree).toMatchSnapshot()
})
