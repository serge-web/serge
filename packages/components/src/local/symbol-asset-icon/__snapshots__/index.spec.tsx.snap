// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly SymbolAssetIcon 1`] = `
<div
  className="asset-icon-background"
>
  <div
    className="asset-icon-with-image"
  >
    <img
      className="img"
      src="data:image/svg+xml;base64,PHN2ZyB4bWxucz0iaHR0cDovL3d3dy53My5vcmcvMjAwMC9zdmciIHZlcnNpb249IjEuMiIgYmFzZVByb2ZpbGU9InRpbnkiIHdpZHRoPSI4IiBoZWlnaHQ9IjgiIHZpZXdCb3g9Ijk2IDk2IDggOCI+PC9zdmc+"
    />
    <span>
      
    </span>
  </div>
</div>
`;
